### PR TITLE
Refactor retry strategy and host management

### PR DESCRIPTION
**Pull request by julienbourdeau**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | no
| New feature?      | yes    
| BC breaks?        | Not sure  
| Related Issue     | Fix #...  
| Need Doc update   | no

## Describe your change

I have refactored the host management to implement the 5minutes timeout if an host is down and to reconnect to the main host as fast as possible.

Next, I'm going to add the cache on the file system to restore the state.

### Help

We need a better name for `ClusterHosts`.

### CODE REVIEW

File: src/RetryStrategy/ClusterHosts.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace Algolia\AlgoliaSearch\RetryStrategy;
+
+class ClusterHosts
```
Comment (nunomaduro) [inline src/RetryStrategy/ClusterHosts.php]: Shouldn't this class marked as final and internal?

File: src/RetryStrategy/Host.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace Algolia\AlgoliaSearch\RetryStrategy;
+
+class Host
```
Comment (nunomaduro) [inline src/RetryStrategy/Host.php]: Shouldn't this class marked as final and internal?

**Comment by julienbourdeau** [comment]
I'll upfate the PR for the conflicts 👍 

About `@internal` I wasn't entirely sure yet what to tag. I think it should be Host and HostCollection. About ClusterHost, I'll check :)

**Comment by julienbourdeau** [comment]
We 'll add `@internals` and `final` when we review

**Pull request merged**